### ISSUE TITLE
Hang while waiting for PhantomJS to die

### ISSUE THREAD

**Issue by lygaret** [teampoltergeist/poltergeist#747]
I've got a bunch of workers in Resque, and somewhat randomly, these workers end up hanging in poltergeist code. I don't have a good explanation, nor specific reproduction steps, but I was hoping someone might have run into this before.

This is the stack trace I obtained when connecting to the ruby process with GDB on my production servers. (Note that the process should have been complete in < 1min, this stack trace was obtained 22 hours after the process started.)

```
/home/scout/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/timeout.rb:93:in `join'
/home/scout/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/timeout.rb:93:in `ensure in block in timeout'
/home/scout/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/timeout.rb:93:in `block in timeout'
/home/scout/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/timeout.rb:34:in `block in catch'
/home/scout/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/timeout.rb:34:in `catch'
/home/scout/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/timeout.rb:34:in `catch'
/home/scout/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/timeout.rb:104:in `timeout'
/home/scout/.rvm/gems/ruby-2.2.0@mvp/gems/poltergeist-1.7.0/lib/capybara/poltergeist/client.rb:118:in `kill_phantomjs'
/home/scout/.rvm/gems/ruby-2.2.0@mvp/gems/poltergeist-1.7.0/lib/capybara/poltergeist/client.rb:72:in `stop'
/home/scout/.rvm/gems/ruby-2.2.0@mvp/gems/poltergeist-1.7.0/lib/capybara/poltergeist/driver.rb:80:in `quit'
/home/scout/MVP/lib/scraper/scrape_helpers.rb:17:in `fetch_live'
[... continues into my application code ...]
```

Having walked through the code, I'm not sure what could be going wrong: it looks like there's a race condition in waiting for the phantomJS's pid (`Process.wait(pid)`) and the Timeout block, but I'm not entirely sure. Has anyone ever seen anything like this?

_Ruby Version_: 2.2.0
_Poltergeist Gem Version_: 1.7.0

**Comment by twalpole** [comment]
So the line your stack trace shows it to be hanging on is

    Timeout.timeout(KILL_TIMEOUT) { Process.wait(pid) }

which shouldn't be able to hang unless Process.wait is non interruptible (which I don't think it is).  I would understand if it was hanging on line 121 which is just Process.wait(pid).  What platform are your production servers running?

**Comment by lygaret** [comment]
> What platform are your production servers running?
Debian Jesse

> shows it to be hanging on...
Well, more explicitly, it's hanging in Timeout#timeout, on the join in these lines:

```
          y.kill
          y.join # make sure y is dead.
```
Which, as far as I can tell shouldn't be possible. So, I have no real idea what's going

**Comment by lygaret** [comment]
In the meantime, I'm getting around it at the moment with the following snippet:

```ruby
                 [... continue ...]
                 session = Capybara::Session.new(:poltergeist)

                 # without this, the timeout thread in kill phantomjs was hanging for some reason
                 # redefine to not do the timeout, and just club the phantomjs process over the head
                 # TODO: HACK: WTF: UGH
                 metasession = class << session.driver.client; self; end
                 metasession.send(:define_method, :kill_phantomjs, proc do
                                    begin
                                      Process.kill('KILL', pid)
                                    rescue => ex
                                      Rails.logger.error(ex, "Error killing phantomjs, #{ex.message}")
                                    end
                                  end)
                 # END OF HACKERY

                 session.visit(url)
                [... continue ...]
```

but I really, REALLY don't like that.

**Comment by twalpole** [comment]
As a guess I would say there's some of race condition happening with the Timeout exception being raised while the kill or join are executing  --  timeout.rb just isn't fully safe for use -- not sure what iother options we have though.

**Comment by twalpole** [comment]
Maybe we could pass Process::WNOHANG to wait and then implement our own loop for timeout - I can't find anything about exactly what platform support there is for WNOHANG though -  @route any other thoughts?

**Comment by brucek** [comment]
I have the same issue on Heroku FWIW.

Ruby Version: 2.3.3
Poltergeist Gem Version:  1.11.0
Heroku Stack:  cedar-14
OS:  Ubuntu 14.04

**Comment by route** [comment]
@twalpole @lygaret who raises that error? How did you get this backtrace? Looks like `y.join` should never ever raise an error because it raises error in `x` thread which is main thread. So it's either should be locked on join or success and return thread but not an error but in the output it's the last point of backtrace

**Comment by lygaret** [comment]
@route

It's been close to a year, so apologies for fuzzy memory, but the stack trace was obtained by attaching GDB to the hung ruby process and getting it from there. 

See: https://blog.newrelic.com/2013/04/29/debugging-stuck-ruby-processes-what-to-do-before-you-kill-9

**Comment by lygaret** [comment]
So, you're right, it didn't throw, just hung, and I generated the stack trace live.

**Comment by route** [comment]
@lygaret @twalpole sorry guys I just didn't notice you mentioned me among all notifications. So it hung on `y.join` in the `ensure` block w/o raise. There are 2 cases we could get into `ensure`:

1) `Process.wait` didn't return on time (hung?)
  a) `y` tries to raise an exception by timeout but it should raise it in `x` thread which hung on `wait`, then we will never go to `ensure`.
  b) `y` raises an exception in `x` thread a while after `wait` returns then we go to `ensure` and I see no reason it could hang on `y.kill` or `y.join` because `y` in fact is already dead.

2) `Process.wait` succeeded in collecting dead bones of a process and we went to `ensure` block trying to call `y.kill` and then `y.join`.
  a) `y` hasn't raised an error and all should be fine because we just kill and join it
  b) `y` tries to raise an error in `x` no matter `yield` succeeded (it's just sleep - raise right?) and the only way it can hang again `x` thread dead locked.

What `kill` does as I can see it just puts one more event into pending interruptions queue https://github.com/ruby/ruby/blob/bc855ef1391bb79ca3554f67ae2d3aad19178d94/thread.c#L2257 it should never hang on kill.

So in both scenarios the only way it can hang is when `x` is dead locked. I see no reason why `y` could be locked. I thought resque https://github.com/resque/resque/search?utf8=%E2%9C%93&q=Thread doesn't use threads but in fact it uses some. `Process::WNOHANG` is no help I think there must be a child process or if there's no ruby raises an exception.

@lygaret @brucek do you use `COUNT` for rescue or threads in the worker?

**Comment by lygaret** [comment]
@route I don't use `COUNT`, at least directly, as I'm managing workers with `resque-pool`, and as far as I can tell it doesn't use `COUNT` but rather manages the workers directly. 

As per threads:  I do use threads in the worker - I've got a thread pool that I'm starting up and then draining, but the threads don't share anything afaik, and the code that's using the poltergeist is pretty straightforward. This is the code I've got:

```ruby
      # scrape the given url with a live DOM
      # yields a parsed DOM (nokogiri document) after the page has settled (ajax on load is complete)
      # uses phantomjs via poltergeist/capybara
      def fetch_live(url, debounce = 0.5)
        raise "cannot fetch a blank URL" unless url.present?
        html = begin
                 session = Capybara::Session.new(:poltergeist)

                 # without this, the timeout thread in kill phantomjs was hanging for some reason
                 # redefine to not do the timeout, and just club the phantomjs process over the head
                 # TODO: HACK: Why does this happen?!
                 metasession = class << session.driver.client; self; end
                 metasession.send(:define_method, :kill_phantomjs, proc do
                                    begin
                                      Process.kill('KILL', pid)
                                    rescue => ex
                                      Rails.logger.error(ex, "Error killing phantomjs, #{ex.message}")
                                    end
                                  end)
                 # END OF HACKERY

                 session.visit(url)
                 sleep debounce if debounce.present?
                 session.html
               ensure
                 session.driver.quit
               end

        document = Nokogiri::HTML(html)
        if block_given?
          yield document
        else
          document
        end
      end
```

Note that the "TODO" block is dropping the timeout which keeps the hang from occuring. That may be called from a not-main thread, but doesn't do it's own threading.

Happy to help with whatever, but we're starting to stretch my memory to the point where I'd have to get back into this for real ;)

Thanks for taking a look!

**Comment by lygaret** [comment]
Other potentially useful info:

* this was happening on a server that's pretty busy, especially around forking processes. We run thousands of these jobs an hour.
* it's really inconsistent, and I'd say it happens < 0.1% of the time, but since we're running so many jobs, my "hung process reaper" was catching about 10 of these a day. 
* the override to remove the timeout cleared these out completely.

So basically, I've got this working for me in the moment, and it happens so infrequently (overall) to be an annoyance for me right now, vs. a showstopper.

**Comment by route** [comment]
@lygaret I would appreciate code around `fetch_live` as I see no threads here and this issues seems to be related to threads...

**Comment by brucek** [comment]
Hi @route, I am also using resque-pool with no additional threading.  We're in limited testing on our code, and I would see this up to several times a day on a server with only 10's of runs - much more frequently than @lygaret (I'd guess 1-10% of runs).  I'm using his `process.kill` hack (thanks Jonathan!) and have also not seen the issue since, FWIW.

**Comment by route** [comment]
Hey @brucek, how does it behave for you without this hack? a worker hangs?

**Comment by brucek** [comment]
Yes, without @lygaret's hack it hangs a lot (as mentioned, maybe during 1-10% of our runs)

**Comment by brucek** [comment]
Following up for random googlers, poltergeist 12.0 users can now monkeypatch with this:

```ruby
module Capybara::Poltergeist
  class Client
    def self.process_killer(pid)
      proc do
        begin
          Process.kill('KILL', pid)
        rescue Errno::ESRCH, Errno::ECHILD
          # Zed's dead, baby
        rescue => ex
          puts "Error killing phantomjs, #{ex.message}"
          raise
        end
      end
    end
  end
end
```

**Comment by twalpole** [comment]
Would anyone be able to try the `hang_to_die` branch and see if it makes a difference?

**Comment by twalpole** [comment]
Closing due to lack of response - we can reopen if/when info is provided.

**Comment by fbernier** [comment]
@twalpole I am experiencing the very same issue described by others. Hanging in resque tasks. I'll see if I can try your branch.

**Comment by fbernier** [comment]
@twalpole I confirm the branch works for us in production. No hang since then.
